core/modules/replica/Job.cc

```diff
@@ -0,0 +1,435 @@
+
+    startImpl(lock);
+
+    // Allow the job to be fully accomplished right away
+
+    if (state() == State::FINISHED) {
+        notify();
+        return;
+    }
+
+    // Otherwise, the only other state which is allowed here is this
+
+    assertState(lock,
+                State::IN_PROGRESS,
+                context() + "start");
+}
+
+void Job::cancel() {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "cancel"
+         << "  state="         << state2string(state())
+         << ", extendedState=" << state2string(extendedState()));
+
+    // IMPORTANT: the final state is required to be tested twice. The first time
+    // it's done in order to avoid deadlock on the "in-flight" requests reporting
+    // their completion while the job termination is in a progress. And the second
+    // test is made after acquering the lock to recheck the state in case if it
+    // has transitioned while acquering the lock.
+
+    if (state() == State::FINISHED) return;
+
+    util::Lock lock(_mtx, context() + "cancel");
+
+    if (state() == State::FINISHED) return;
+
+    finish(lock, ExtendedState::CANCELLED);
+}
+
+void Job::finish(util::Lock const& lock,
+                 ExtendedState newExtendedState) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "finish"
+         << "  state="              << state2string(state())
+         << ", extendedState="      << state2string(extendedState())
+         << ", (new)extendedState=" << state2string(newExtendedState));
+
+    // Also ignore this event if the request is over
+
+    if (state() == State::FINISHED) return;
+
+    // *IMPORTANT*: Set new state *BEFORE* calling subclass-specific cancellation
+    // protocol to make sure all event handlers will recognize this scenario and
+    // avoid making any modifications to the request's state.
+
+    setState(lock,
+             State::FINISHED,
+             newExtendedState);
+
+    // Invoke a subclass specific cancellation sequence of actions if anything
+    // bad has happen.
+
+    if (newExtendedState != ExtendedState::SUCCESS) cancelImpl(lock);
+
+    controller()->serviceProvider()->databaseServices()->saveState(*this);
+
+    // Stop timers if they're still running
+
+    if(_heartbeatTimerPtr)  _heartbeatTimerPtr->cancel();
+    if(_expirationTimerPtr) _expirationTimerPtr->cancel();
+
+    notify();
+}
+
+void Job::qservAddReplica(util::Lock const& lock,
+                          unsigned int chunk,
+                          std::vector<std::string> const& databases,
+                          std::string const& worker,
+                          AddReplicaQservMgtRequest::CallbackType onFinish) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context()
+         << "** START ** Qserv notification on ADD replica:"
+         << ", chunk="     << chunk
+         << ", databases=" << util::printable(databases)
+         << "  worker="    << worker);
+
+    auto self = shared_from_this();
+
+    controller()->serviceProvider()->qservMgtServices()->addReplica(
```

Comment:
never null! :-)

Reply:
Now you're arguing with @andy-slac because what you're seeing is a result of his request to migrate from a shared_ptr to a reference :-)

Reply:
...and it's never going to be 0